gcp_variant_transforms/general_process.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2018 Google Inc.  All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""General functions that are used by both vcf_to_bq and preprocessor.
+
+It includes parsing the command line arguments, reading the input, applying the
+PTransforms and writing the output.
+"""
+
+from argparse import ArgumentParser
+from argparse import Namespace  # pylint: disable=unused-import
+from typing import List  # pylint: disable=unused-import
+import enum
+
+import apache_beam as beam
+from apache_beam.io.filesystems import FileSystems
+from apache_beam import Pipeline  # pylint: disable=unused-import
+from apache_beam.pvalue import PCollection  # pylint: disable=unused-import
+
+from gcp_variant_transforms.beam_io import vcfio
+from gcp_variant_transforms.beam_io import vcf_header_io
+from gcp_variant_transforms.options.variant_transform_options import VariantTransformsOptions  # pylint: disable=unused-import
+from gcp_variant_transforms.transforms import filter_variants
+from gcp_variant_transforms.transforms import infer_undefined_headers
+from gcp_variant_transforms.transforms.merge_headers import MergeHeaders  # pylint: disable=unused-import
+
+# If the # of files matching the input file_pattern exceeds this value, then
+# headers will be merged in beam.
+_SMALL_DATA_THRESHOLD = 100
+_LARGE_DATA_THRESHOLD = 50000
+
+
+class PipelineModes(enum.Enum):
+  """An Enum specifying the mode of the pipeline based on the data size."""
+  SMALL = 0
+  MEDIUM = 1
+  LARGE = 2
+
+
+def parse_args(argv, command_line_options):
+  # type: (List[str], List[type]) -> (Namespace, List[str])
+  """Parses the arguments.
+
+  Args:
+    argv: A list of string representing the pipeline arguments.
+    command_line_options: A list of type ``VariantTransformsOptions`` that
+      specifies the options that will be added to parser.
+  """
+  parser = ArgumentParser()
+  parser.register('type', 'bool', lambda v: v.lower() == 'true')
+  options = [option() for option in command_line_options]
+  _add_parser_arguments(options, parser)
+  known_args, pipeline_args = parser.parse_known_args(argv)
+  _validate_args(options, known_args)
+  return known_args, pipeline_args
+
+
+def get_pipeline_mode(known_args):
+  # type: (Namespace) -> int
+  """Returns the mode the pipeline should operate in based on input size."""
+  if known_args.optimize_for_large_inputs:
+    return PipelineModes.LARGE
+
+  match_results = FileSystems.match([known_args.input_pattern])
```

Comment:
@bashir2 For the type hints `Namespace`, I got warnings as unresolved attribute reference for class `Namespace`, I didn't find a good way to resolve it. Any suggestions?

Reply:
Right, the type checking system cannot understand these kind of dynamic objects (thanks to Python). With a quick look at argparse, I think these attributes (e.g., input_pattern) are added dynamically.

BTW, you can just use argparse.Namespace if you want to avoid importing the Namespace class (since the style guide recommends against it).